arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -322,36 +351,114 @@ static bool sync_shadow_context_entry(struct pgt_sync_data *sdata)
 {
 	struct context_entry *shadow_ce = sdata->shadow_ptep, tmp = {0};
 	struct context_entry *guest_ce = sdata->guest_ptep;
+	struct pkvm_iommu *iommu = pgt_to_pkvm_iommu(sdata->spgt);
+	struct pkvm_ptdev *ptdev;
+	struct pkvm_pgtable_cap cap;
 	bool updated = false;
-	u8 aw;
+	u8 tt, aw;
+	u16 bdf;
 
-	if (sdata->guest_ptep && sdata->shadow_pa) {
-		tmp.hi = guest_ce->hi;
-		tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
+	if (ecap_smts(sdata->iommu_ecap)) {
+		if (sdata->guest_ptep && sdata->shadow_pa) {
+			tmp.hi = guest_ce->hi;
+			tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
 
-		if (ecap_smts(sdata->iommu_ecap))
 			/* Clear DTE to make sure device TLB is disabled for security */
 			context_sm_clear_dte(&tmp);
-		else {
+		}
+	} else {
+		/*
+		 * In legacy mode, a context entry is a leaf entry responsible for
+		 * configuring the actual address translation for the given ptdev,
+		 * much like a PASID table entry in scalable mode. So the below logic
+		 * is quite similar to the logic in sync_shadow_pasid_table_entry()
+		 * for scalable mode.
+		 */
+		bdf = sdata->vaddr >> LM_DEVFN_SHIFT;
+		ptdev = iommu_find_ptdev(iommu, bdf, 0);
+
+		if (!ptdev) {
+			ptdev = iommu_add_ptdev(iommu, bdf, 0);
+			if (!ptdev)
+				return false;
+		}
+
+		if (!sdata->guest_ptep) {
+			if (context_lm_is_present(shadow_ce)) {
+				pkvm_setup_ptdev_vpgt(ptdev, 0, NULL, NULL, NULL);
+				pkvm_setup_ptdev_did(ptdev, 0);
+				iommu_del_ptdev(iommu, ptdev);
+
+				goto out;
+			}
+			return false;
+		}
+
+		tt = context_lm_get_tt(guest_ce);
+		switch (tt) {
+		case CONTEXT_TT_MULTI_LEVEL:
+		case CONTEXT_TT_DEV_IOTLB:
+			aw = context_lm_get_aw(guest_ce);
+			if (aw != 1 && aw != 2 && aw != 3) {
+				pkvm_err("pkvm: unsupported address width %u\n", aw);
+
+				pkvm_setup_ptdev_vpgt(ptdev, 0, NULL, NULL, NULL);
+				pkvm_setup_ptdev_did(ptdev, 0);
+				goto out;
+			}
```

Comment:
Hmm, we already have pkvm_err, clearing ptdev info and clearing the shadow entry, what else do we need here?
Regarding scalable mode's flpm, yes, the same sanity check would be good, although IIUC it is not too critical, it would help to detect the host's misbehavior, but pKVM protection is ensured anyway.

Reply:
it's not critical, so I would like to keep a TODO here, could you add some TODO comments here?

Reply:
My point is that I don't understand what this TODO should be about, i.e. what else can we do besides what we already do here? In other words, why do you think this error reporting is not good enough?

Reply:
It's related with error handling emulation, like, how real IOMMU HW behavior when SW set incorrect aw? I suppose it will trigger a translation fault.  Currently we don't have such and just keep silence to host VM.

Reply:
If I understand correct, we already have it: after `goto out` we clear the shadow context entry, in particular its "present" bit, so it will trigger translation fault.

Reply:
Maybe I should rename this label from `out` to `update_shadow_ce` to make the code more clear?

Reply:
rename the label is good, but here we actually update shadow_ce to 0, the translation fault type is different (context-table fault vs. second-level table fault), as we pass-thru the fault/interrupt to host, maybe we can ignore the check for aw, just set guest_ce aw to shadow_ce?

Reply:
I've added the TODO here. I've also noted that we have the same issue in case of incorrect translation type (tt) value.
Why I'm not sure that we can just pass incorrect aw or tt values to the shadow entry: per VT-d spec such values are actually "reserved", so in future HW they may mean some correct translation setup, and we don't know how to handle such future correct setup in pKVM to ensure protection.

Reply:
I think we are actually pass-thru aw setting, but tt is not. And you are right, for "reserved" we cannot foresee what's the future usage. I have to say the most correct way to do the emulation is check the value and do fault injection based on it. But it seemed not easy now, so I may still prefer to allow pass-thru aw setting (but still add TODO here)